src/components/implementation/tests/micro_llping/ping.c

```diff
@@ -1231,101 +1450,89 @@ void cos_init(void)
 	/* } */
 
 	arcv_ready[cos_cpuid()] = 1;
-#if NUM_CPU > 2
-//	if (cos_cpuid() < (NUM_CPU_COS - SND_RCV_OFFSET)  && (cos_cpuid() % 4 == 0)) {
-//	if ((cos_cpuid()%4 == 0 || cos_cpuid()%4 == 2) && (cos_cpuid()+SND_RCV_OFFSET < NUM_CPU_COS)) { // sending core
-	if (cos_cpuid() <= 0) {
-//	if (1) {
-		/* IPI - ASND/ARCV */
-
-//		ipi_test();
-//		cap_test();
-//		mem_test();
-//		thd_test();
-//		wcet_test();
-//		cons_decons_test();
-//		retype_test();
-//		kobj_test();
-//		response_test();
-		captbl_cons_test();
-	} else { //if ((cos_cpuid() % 4 <= 1)
-//		printc("core %ld: thd %d switching to pong thd\n", cos_cpuid(), cos_get_thd_id());
-		while (1) {
-			//do op here to measure response time.
-			/* call_cap(4, 0, 0, 0, 0); */
-			/* rdtscll(e); */
-			/* if ((e-s)/(2000*1000*1000) > RUNTIME) break; */
+/* #if NUM_CPU <= 2 */
+/* 	pingpong(); */
+/* #else */
+	int ret, cpu = cos_cpuid(); 
+	unsigned long i;
+	unsigned long long s,e, tot = 0, tot2 = 0;
 
-			if (ck_pr_load_int(&all_exit)) break;
-		}
-		printc("core %ld: exiting from ping\n", cos_cpuid());
+	if (cos_cpuid() == 0) {
+		printc(">>>>> calling mm init\n");
+		call_cap(4, 0, 0, 0, 0);
+		printc(">>>>> done mm init\n");
+	}
+	sync_all();
+
+	if (mm_meas()) goto done;
+
+	int repeat;
+	unsigned long nops = 0, nops2 = 0;
+
+	play_trace(&nops, &tot, &nops2, &tot2);
+	/* workaround for the liveness id issue */
+	tlb_quiescence_wait();
+	sync_all();
+
+	if (cos_cpuid()) goto done;
+
+	if (cos_cpuid() == 0) nlog = 0;
+
+	nops = tot = 0;
+	nops2 = tot2 = 0;
+	for (repeat = 0; repeat < 10; repeat++) {
+		play_trace(&nops, &tot, &nops2, &tot2);
 	}
+
+	if (nops) {
+		printc("nops %d cpu %d avg cost %llu (%lu ops), avg2 cost %llu (%lu ops)\n", 
+		       NOPS, cpu, (tot)/(nops), nops, tot2/nops2, nops2);
+	} else
+		printc("cpu %d no ops\n", cpu);
+
+	if (cos_cpuid() == 0) {
+		unsigned long j, temp;
+		for (i = 0; i < nlog; i++) {
+			for (j = i + 1; j < nlog; j++) {
+				if (logs[i] < logs[j]) {
+					temp = logs[i];
+					logs[i] = logs[j];
+					logs[j] = temp;
+				}
+			}
+		}
+
+		for (i = 0; i < nlog_large; i++) {
+			for (j = i + 1; j < nlog_large; j++) {
+				if (logs_large[i] < logs_large[j]) {
+					temp = logs_large[i];
+					logs_large[i] = logs_large[j];
+					logs_large[j] = temp;
+				}
+			}
+		}
+
+		printc("nlog %lu on core 0, SINGLE PAGE 99p (%lu) is %lu , 1024 page (%lu ops) 99p is %lu\n", 
+		       nlog, nlog/100, logs[nlog/100], nlog_large, logs_large[nlog_large/100]);
+#ifdef MEAS_MAP
+		printc("Operation: mmap on %d cores\n", NUM_CPU_COS);
 #else
-	pingpong();
+		printc("Operation: munmap on %d cores\n", NUM_CPU_COS);
 #endif
+	}
+/* #endif */
+done:
+	sync_all();
 	cap_switch_thd(SCHED_CAPTBL_ALPHATHD_BASE + cos_cpuid()*captbl_idsize(CAP_THD));
 
+	/* help linking. hack for now. */
+	mman_get_page(0,0,0);
```

Comment:
Why is this a hack?  Doe this mean that the proper function symbol -> capability -> component mapping is now working?